cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/VerificationStage.java

```diff
@@ -173,6 +155,66 @@ public void process(ApplicationDeployable input) throws Exception {
     emit(input);
   }
 
+  private void verifyWorkflowSpecifications(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec) {
+    List<WorkflowNode> nodes = workflowSpec.getNodes();
+    for (WorkflowNode node : nodes) {
+      verifyWorkflowNode(appSpec, workflowSpec, node);
+    }
+  }
+
+  private void verifyWorkflowNode(ApplicationSpecification appSpec, WorkflowSpecification workflowSpec,
+                                  WorkflowNode node) {
+    WorkflowNodeType nodeType = node.getType();
+    switch (nodeType) {
+      case ACTION:
+        WorkflowActionNode actionNode = (WorkflowActionNode) node;
+        verifyWorkflowAction(appSpec, actionNode.getProgram());
+        break;
+      case FORK:
+        WorkflowForkNode forkNode = (WorkflowForkNode) node;
+        if (forkNode.getBranches().size() == 0) {
+          throw new RuntimeException(String.format("Fork is added in the Workflow '%s' without any branches",
+                                                   workflowSpec.getName()));
+        }
+
+        for (WorkflowForkBranch branch : forkNode.getBranches()) {
```

Comment:
To make switch looks nicer and be parity with the `ACTION` case, better encapsulate this into a method `verifyWorkflowFork`